README.md
# Gitlab-api-generic-client

[![Build Status](https://img.shields.io/travis/emri99/gitlab-generic-api-client/master.svg?style=flat-square)](https://travis-ci.org/emri99/gitlab-generic-api-client)
[![PHP: 5.4](https://img.shields.io/badge/PHP-5.4-blue.svg?style=flat-square)](http://php.net)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square)](https://opensource.org/licenses/MIT)

This library has been built keeping in mind that Gitlab move so fast, that 
it becomes hard to apply changes and migration guides on complex code base.

## How it works

This code is inpired by npm package [gitlab-api-client](https://www.npmjs.com/package/gitlab-api-client).  
Next lines descriptions too.

Main principle: All paths are build generically.  
You aren't stick to any specific API version as you will have access to 
all the gitlab API endpoints, even for those that haven't been defined yet.

## Installation

```
composer require emri99/gitlab-generic-api-client
```

## Usage

### Authentication

* Authenticate using HTTP token
````php
$client->authenticate('SECRET-HTTP-TOKEN', GitlabApiClient::AUTH_HTTP_TOKEN);
````

* Authenticate using OAUTH token
````php
$client->authenticate('SECRET-OAUTH-TOKEN', GitlabApiClient::AUTH_OAUTH_TOKEN);
````

### Requesting

* GET request

````php
$client = new GitlabApiClient('https://my.gitlab.com/api/v4');
$branches = $client->projects(1)
    ->repository()
    ->branches()
    ->get()
    
// will send GET request on 
// https://my.gitlab.com/api/v4/projects/1/repository/branches.

foreach($branches as $branch) {
    echo $branch->name, "\n";
}
````

* POST request

````php
# create a variable secret
$variableDatas = $this->getClient()
    ->projects(2)
    ->variables()
    ->post([
        'key' => 'SECRET',
        'value' => 'password'
    ]);
````

* PUT request

````php
# protect a branch
$branchUpdated = $this->getClient()
    ->projects(2)
    ->repository()
    ->branches('master')
    ->protect()->put([
        'developers_can_push' => false,
        'developers_can_merge' => false
    ]);
$done = $branchUpdated->protected;
````

* DELETE request

````php
# delete a branch
$branchUpdated = $this->getClient()
    ->projects(2)
    ->repository()
    ->branches('obsolet-feature')
    ->delete();
````

### Special case

If an url segment is the same than a public method of `GitlabApiClient`, this 
remains possible to build to path correctly.

For example, to build path `user/1/delete`, use:

```
$client->user(1, 'delete');
```

### IDE Completion depending on gitlab api version (optional)

Empty classes can be used to simulate code completion on retrieved object
by installing `emri99/gitlab-generic-api-client-models`.
This [optional package](https://github.com/emri99/gitlab-generic-api-client-models) is tagged by gitlab API version.



*Currently there isn't many versions handled, only the one I'm using. ie: 9.1.4*

> When using this package, retrieved objects **WONT BE** instance of models class.  
> Retrieved objects remains `stdclass`. This is **ONLY** used for IDE completion.

```
composer require emri99/gitlab-generic-api-client-models:YOUR_GITLAB_VERSION --dev
```

YOU **MUST** add phpdoc to use completion like below:

* **GET**
````php
$client = new GitlabApiClient('https://my.gitlab.com/api/v4');

/** 
 * $branches aint really a Branch instance 
 * @var Branch[] $branches 
 */
$branches = $client->projects(1)
    ->repository()
    ->branches()
    ->get(array(
        // parameters
    ))

// $branches is an array of stdclass having 
// the same properties than a Branch class
foreach($branches as $branch) {
    echo $branch->name;
}
````

# Contributing

Thanks for contributing !

Please follow this rules:
* you MUST apply supplied CS-fixer by running `composer run-script cs`
* you MUST write/update the tests
* you SHOULD write documentation
* you MUST write minimum details in pull request description

Squashing many commits to avoid noise on git logs make sense too ;)


CHANGELOG.md
## v3.0.0
* styte: apply CS fixer
* build(composer): update composer run scripts
* build(composer): Add dev-master alias on 3.0.x-dev
* feat(error): Add Exception classes
  > 404 errors now throws a `\Emri99\Gitlab\Exception\NotFoundException`

**BC changes:**
* `\RuntimeException` have been replaced with `\Emri99\Gitlab\Exception\GitlabApiClientException`

## v2.0.3
fix: detection of empty response
fix: remove PHP notice warning on missing Content-Type response header

## v2.0.2
fix: Allow usage of null parameter to pass optional url argument

Previously, if arguments pass as `$api->projects($idOrName)` were failing if `$idOrName` was null, expecting a scalar type, now ignore arguments if it's value is null.

This allow to use `$api->projects($idOrName)`:
* to request `/projects` when `$idOrName` is null
* to request `/projects/{idOrName}` otherwise

## v2.0.1
* fix: POST/PUT requests handling
  * use correct content-type `x-www-form-urlencoded`or `multipart/form-data` when files supplied
* refacto: GET/DELETE requests handling
* remove: PATCH request (no used on gitlab api currently)
* improvement: response error handling
* test: update test

## v2.0.0
* Throw Exception on response status code >= 400

## v1.0.0
* Initial version




tests/bootstrap.php
<?php

/*
 * This file is part of emri99/gitlab-generic-api-client.
 *
 * (c) 2017 Cyril MERY <mery.cyril@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

$loader = require __DIR__.'/../vendor/autoload.php';
$loader->add('Emri99\\Gitlab\\', __DIR__);


tests/GitlabApiClientTest.php
<?php

/*
 * This file is part of emri99/gitlab-generic-api-client.
 *
 * (c) 2017 Cyril MERY <mery.cyril@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Emri99\Gitlab;

use Emri99\Gitlab\GitlabApiClient as SUT;
use Unirest\Response;

/**
 * @covers \Emri99\Gitlab\GitlabApiClient
 */
class GitlabApiClientTest extends \PHPUnit_Framework_TestCase
{
    const BASE_PATH = 'http://example.com';

    public function testConstants()
    {
        $this->assertEquals('HTTP', SUT::AUTH_HTTP_TOKEN);
        $this->assertEquals('OAUTH', SUT::AUTH_OAUTH_TOKEN);
    }

    public function testBuildComplexPath()
    {
        $sut = $this->getMockedSUT();
        $sut
            ->expects($this->once())
            ->method('performRequest')
                ->with(
                    'GET',
                    self::BASE_PATH.'/path/1/to/data%20one/for/test?param2=value2',
                    null,
                    $this->isType('array'),
                    $this->isType('array')
                )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut->path(1)->to('data one')->for('test')->get(array(
            'param2' => 'value2',
        ));
    }

    public function testAuthenticateMissingMode()
    {
        $sut = $this->getMockedSUT();
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'PUT',
                self::BASE_PATH.'/path',
                $this->anything(),
                $this->isEmpty(),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate('wont-be-passed', 'inexistent-mode')
            ->path()
            ->put();
    }

    public function testAuthenticateHttpToken()
    {
        $token = 'my-precious';
        $sudo = 'special sudo string';
        $self = $this; // for closure

        $sut = $this->getMockedSUT(self::BASE_PATH.'/');
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'PUT',
                self::BASE_PATH.'/path/to/put',
                $this->anything(),
                $this->callback(function ($datas) use ($self) {
                    $self->assertEquals(array('PRIVATE-TOKEN' => 'my-precious',
                    ), $datas);

                    return true;
                }),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate($token, SUT::AUTH_HTTP_TOKEN)
            ->path('to', 'put')
            ->put();

        $sut = $this->getMockedSUT();
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'POST',
                self::BASE_PATH.'/path/2/to/post',
                $this->anything(),
                $this->callback(function ($datas) use ($self) {
                    $self->assertEquals(array(
                        'PRIVATE-TOKEN' => 'my-precious',
                        'SUDO' => 'special sudo string',
                    ), $datas);

                    return true;
                }),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate($token, SUT::AUTH_HTTP_TOKEN, $sudo)
            ->path(2)
            ->to('post')
            ->post();
    }

    public function testAuthenticateOauthToken()
    {
        $token = 'my-precious';
        $sudo = 'special sudo string';
        $self = $this;

        $sut = $this->getMockedSUT();
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'PUT',
                self::BASE_PATH.'/path/to/put',
                $this->anything(),
                $this->callback(function ($datas) use ($self) {
                    $self->assertEquals(array('Authorization' => 'Bearer my-precious',
                    ), $datas);

                    return true;
                }),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate($token, SUT::AUTH_OAUTH_TOKEN)
            ->path('to', 'put')
            ->put();

        $sut = $this->getMockedSUT();
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'DELETE',
                self::BASE_PATH.'/path/2/to/delete',
                $this->anything(),
                $this->callback(function ($datas) use ($self) {
                    $self->assertEquals(array(
                        'Authorization' => 'Bearer my-precious',
                        'SUDO' => 'special sudo string',
                    ), $datas);

                    return true;
                }),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate($token, SUT::AUTH_OAUTH_TOKEN, $sudo)
            ->path(2)
            ->to('delete')
            ->delete();
    }

    /**
     * @expectedExceptionMessage Page not found : http://example.com/path
     * @expectedException \Emri99\Gitlab\Exception\NotFoundException
     */
    public function testDisplayUrlOn404()
    {
        $sut = $this->getMockedSUT();
        $sut
            ->expects($this->once())
            ->method('performRequest')
            ->with(
                'GET',
                $this->anything(),
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->willReturn(new Response(404, '{}', '', array()));

        $sut->path()->get();
    }

    /**
     * @expectedExceptionMessage No response
     * @expectedException \Emri99\Gitlab\Exception\GitlabApiClientException
     */
    public function testNoResponse()
    {
        $sut = $this->getMockedSUT();
        $sut
            ->expects($this->once())
            ->method('performRequest')
            ->with(
                'GET',
                $this->anything(),
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->willReturn(null);

        $sut->path()->get();
    }

    /**
     * @expectedExceptionMessage Unable to decode json response
     * @expectedException \Emri99\Gitlab\Exception\GitlabApiClientException
     */
    public function testBadResponse()
    {
        $sut = $this->getMockedSUT();
        $sut
            ->expects($this->once())
            ->method('performRequest')
            ->with(
                'GET',
                $this->anything(),
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->willReturn(new Response(200, '{malformed json}', 'Content-Type: application/json', array()));

        $sut->path()->get();
    }

    public function testOptions()
    {
        $client = new SUT(self::BASE_PATH, array(
            'new_option' => 'value',
            'timeout' => 666,
        ));
        $this->assertEquals('value', $client->getOption('new_option'));
        $this->assertEquals(666, $client->getOption('timeout'));
    }

    /**
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidOptions()
    {
        $client = new SUT(self::BASE_PATH);
        $client->getOption('missing-option');
    }

    public function testHeaders()
    {
        $self = $this;

        $sut = $this->getMockedSUT();
        $sut->expects($this->exactly(2))
            ->method('performRequest')
            ->withConsecutive(
                // first call have all headers
                // * globally defined ones
                // * current request ones
                array(
                    'GET',
                    $this->anything(),
                    $this->anything(),
                    $this->callback(function ($datas) use ($self) {
                        $self->assertEquals(array(
                            'HEADER1' => 'VALUE1',
                            'HEADER2' => 'VALUE2',
                            'HEADER3' => 'VALUE3',
                            'HEADER4' => 'VALUE4',
                        ), $datas);

                        return true;
                    }),
                    $this->anything(),
                ),
                // next call must have only global ones
                array(
                    'GET',
                    $this->anything(),
                    $this->anything(),
                    $this->callback(function ($datas) use ($self) {
                        $self->assertEquals(array(
                            'HEADER1' => 'VALUE1',
                            'HEADER2' => 'VALUE2',
                        ), $datas);

                        return true;
                    }),
                    $this->anything(),
                )
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->setHeaders(array(
                'HEADER1' => 'VALUE1',
                'HEADER2' => 'VALUE2',
            ))
            ->path(2)
            ->get(array(), array(
                'HEADER3' => 'VALUE3',
                'HEADER4' => 'VALUE4',
            ));
        $sut->path(2)->get();

        // custom headers must not replace authentication header with same name
        $sut = $this->getMockedSUT();
        $sut->expects($this->once())
            ->method('performRequest')
            ->with(
                'GET',
                $this->anything(),
                $this->anything(),
                $this->callback(function ($datas) use ($self) {
                    $self->assertEquals(array(
                        'HEADER2' => 'VALUE2',
                        'HEADER3' => 'VALUE3',
                        'HEADER4' => 'VALUE4',
                        'Authorization' => 'Bearer my-precious-token',
                        'SUDO' => 'im sudoers',
                    ), $datas);

                    return true;
                }),
                $this->anything()
            )
            ->willReturn(new Response(200, '{}', '', array()));

        $sut
            ->authenticate('my-precious-token', SUT::AUTH_OAUTH_TOKEN, 'im sudoers')
            ->setHeaders(array(
                'Authorization' => 'VALUE1',
                'HEADER2' => 'VALUE2',
            ))
            ->path(2)
            ->get(array(), array(
                'HEADER3' => 'VALUE3',
                'HEADER4' => 'VALUE4',
            ));
    }

    /**
     * @expectedExceptionMessage scalar expected
     * @expectedException \InvalidArgumentException
     */
    public function testIncorrectPath()
    {
        $this->getMockedSUT()->path(array('array' => 'not accepted here'));
    }

    protected function getMockedSUT($baseUrl = self::BASE_PATH, $methods = array('performRequest'))
    {
        return $this
            ->getMockBuilder('Emri99\\Gitlab\\GitlabApiClient')
            ->setConstructorArgs(array($baseUrl))
            ->enableOriginalConstructor()
            ->setMethods($methods)
            ->getMock();
    }
}


src/GitlabApiClient.php
<?php

/*
 * This file is part of emri99/gitlab-generic-api-client.
 *
 * (c) 2017 Cyril MERY <mery.cyril@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Emri99\Gitlab;

use Emri99\Gitlab\Exception\GitlabApiClientException;
use Emri99\Gitlab\Exception\NotFoundException;
use Unirest\Method;
use Unirest\Request;
use Unirest\Response;

class GitlabApiClient
{
    /**
     * Authenticate mode : HTTP Header
     * Use an personal/impersonate access token.
     */
    const AUTH_HTTP_TOKEN = 'HTTP';

    /**
     * Authenticate mode : OAUTH token
     * Use the key supplied by Gitlab's OAuth provider.
     */
    const AUTH_OAUTH_TOKEN = 'OAUTH';

    /**
     * Dynamic part of url.
     *
     * @var string
     */
    protected $path;

    /**
     * Default headers (supplied).
     *
     * @var array
     */
    protected $headers = array();

    /**
     * Authentication headers (computed internally).
     *
     * @var array
     */
    protected $authenticationHeaders = array();

    /**
     * @var array|null
     */
    protected $auth = null;

    /**
     * @var array
     */
    private $options = array();

    private $baseUrl;

    public function __construct($baseUrl, array $options = array())
    {
        if ('/' === substr($baseUrl, -1, 1)) {
            $baseUrl = substr($baseUrl, 0, -1);
        }
        $this->baseUrl = $baseUrl;
        $this->options = array_merge(array(
            'user_agent' => 'php-gitlab-generic-api-client',
            'timeout' => 60,
            'json_decode_to_array' => false,
            'json_decode_options' => JSON_NUMERIC_CHECK & JSON_FORCE_OBJECT & JSON_UNESCAPED_SLASHES & JSON_UNESCAPED_UNICODE,
            'verify_peer' => true,
            'verify_host' => true,
        ), $options);
    }

    /**
     * Add url segment to endpoint url.
     *
     * @param string $method
     *
     * @return $this
     */
    public function __call($method, array $args)
    {
        $this->path .= '/'.$method;

        if (0 === count($args)) {
            return $this;
        }

        foreach ($args as $param) {
            if (null === $param) {
                continue;
            }

            if (!is_scalar($param)) {
                throw new \InvalidArgumentException(sprintf('Method "%s" : Invalid argument type. scalar expected.', $method));
            }
            $this->__call($this->encodePath($param), array());
        }

        return $this;
    }

    /**
     * @return $this
     */
    public function setHeaders(array $headers)
    {
        $this->headers = $headers;

        return $this;
    }

    /**
     * @param string $name
     *
     * @throws \InvalidArgumentException
     *
     * @return mixed
     */
    public function getOption($name)
    {
        if (!array_key_exists($name, $this->options)) {
            throw new \InvalidArgumentException(sprintf('Undefined option called: "%s"', $name));
        }

        return $this->options[$name];
    }

    /**
     * Authenticate a user for all next requests.
     *
     * @param string $token      Gitlab private token
     * @param string $authMethod One of the AUTH_* class constants
     * @param string $sudo
     *
     * @return $this
     */
    public function authenticate($token, $authMethod = self::AUTH_HTTP_TOKEN, $sudo = null)
    {
        $this->authenticationHeaders = array();

        if (!in_array($authMethod, array(self::AUTH_HTTP_TOKEN, self::AUTH_OAUTH_TOKEN))) {
            return $this;
        }

        if (self::AUTH_HTTP_TOKEN === $authMethod) {
            $this->authenticationHeaders['PRIVATE-TOKEN'] = $token;
        }

        if (self::AUTH_OAUTH_TOKEN === $authMethod) {
            $this->authenticationHeaders['Authorization'] = sprintf('Bearer %s', $token);
        }

        if (!is_null($sudo)) {
            $this->authenticationHeaders['SUDO'] = $sudo;
        }

        return $this;
    }

    /**
     * @param array $headers
     *
     * @throws GitlabApiClientException
     * @throws NotFoundException
     *
     * @return mixed
     */
    public function get(array $parameters = array(), $headers = array())
    {
        $response = $this->request(Method::GET, $this->flushPath($parameters), $headers);

        return $this->returnOrThrow($response);
    }

    /**
     * @param array $headers
     *
     * @throws GitlabApiClientException
     * @throws NotFoundException
     *
     * @return mixed
     */
    public function post(array $parameters = array(), $headers = array(), array $files = array())
    {
        $body = null;
        if (!empty($parameters) && empty($files)) {
            $headers['Content-Type'] = 'application/x-www-form-urlencoded';
            $body = Request\Body::Form($parameters);
        } elseif (!empty($files)) {
            $headers['Content-Type'] = 'multipart/form-data';
            $body = Request\Body::Multipart($parameters, $files);
        }

        $response = $this->request(Method::POST, $this->flushPath(), $headers, $body);

        return $this->returnOrThrow($response);
    }

    /**
     * @param array $headers
     *
     * @throws GitlabApiClientException
     * @throws NotFoundException
     *
     * @return mixed
     */
    public function put(array $parameters = array(), $headers = array(), array $files = array())
    {
        $body = null;
        if (!empty($parameters) && empty($files)) {
            $headers['Content-Type'] = 'application/x-www-form-urlencoded';
            $body = Request\Body::Form($parameters);
        } elseif (!empty($files)) {
            $headers['Content-Type'] = 'multipart/form-data';
            $body = Request\Body::Multipart($parameters, $files);
        }

        $response = $this->request(Method::PUT, $this->flushPath(), $headers, $body);

        return $this->returnOrThrow($response);
    }

    /**
     * @param array $headers
     *
     * @throws GitlabApiClientException
     * @throws NotFoundException
     *
     * @return mixed
     */
    public function delete(array $parameters = array(), $headers = array())
    {
        $response = $this->request('DELETE', $this->flushPath($parameters), $headers);

        return $this->returnOrThrow($response);
    }

    /**
     * @param Response $response
     *
     * @throws NotFoundException
     * @throws GitlabApiClientException
     *
     * @return array|\stdClass depending on option 'json_decode_to_array'
     */
    protected function returnOrThrow($response)
    {
        if (!$response) {
            throw new GitlabApiClientException('No response');
        }

        if (null === $response->raw_body || '' === $response->raw_body) {
            return null;
        }

        if (is_string($response->body)
            && isset($response->headers['Content-Type'])
            && 'application/json' === $response->headers['Content-Type']) {
            throw new GitlabApiClientException('Unable to decode json response');
        }

        $error = !empty($response->body->error) ? $response->body->error : null;
        if ($error && !is_string($error)) {
            $error = trim(json_encode($error), '"');
        }

        if ($response->code >= 400) {
            $error = $error ?: '';
            if (404 === $response->code && $this->lastpath) {
                $exception = NotFoundException::create($this->lastpath, $error);
                $this->lastpath = null;
                throw $exception;
            }
        }

        if (null !== $error) {
            throw new GitlabApiClientException(sprintf('%s - %s', $response->code, $error), $response->code);
        }

        if (!$this->options['json_decode_to_array']) {
            return $response->body;
        }

        return json_decode(json_encode($response->body), true);
    }

    protected function request($httpMethod = 'GET', $path, array $headers = array(), $body = null)
    {
        $httpMethod = strtoupper($httpMethod);
        $path = trim($this->baseUrl.$path, '/');

        $httpOptions = array(
            'verifyPeer' => array($this->options['verify_peer']),
            'verifyHost' => array($this->options['verify_host']),
            'timeout' => array($this->options['timeout']),
            'jsonOpts' => array(false, 512, $this->options['json_decode_options']),
        );

        $allHeaders = array_merge($this->headers, $this->authenticationHeaders, $headers);

        return $this->performRequest(
            $httpMethod,
            $path,
            $body,
            $allHeaders,
            $httpOptions
        );
    }

    /**
     * @codeCoverageIgnore
     *
     * @param string $httpMethod
     * @param string $path
     * @param string $body
     * @param array  $allHeaders
     * @param array  $httpOptions
     *
     * @throws \Unirest\Exception
     *
     * @return \Unirest\Response
     */
    protected function performRequest($httpMethod, $path, $body, $allHeaders, $httpOptions)
    {
        foreach ($httpOptions as $method => $args) {
            call_user_func_array(array('Unirest\Request', $method), $args);
        }

        return Request::send($httpMethod, $path, $body, $allHeaders);
    }

    /**
     * @param string $path
     *
     * @return string
     */
    protected function encodePath($path)
    {
        $path = rawurlencode($path);

        return str_replace('.', '%2E', $path);
    }

    /**
     * Retrieve current url fragment & flush it from next requests.
     *
     * @param mixed|null $parameters
     *
     * @return string
     */
    protected function flushPath($parameters = null)
    {
        $path = $this->path;

        if ($parameters) {
            $separator = false === strpos($path, '?') ? '?' : '&';
            $path .= $separator.http_build_query($parameters);
        }

        // remember full url to display path in 404 message
        $this->lastpath = $this->baseUrl.$path;
        $this->path = '';

        return $path;
    }
}
